ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+  # TODO: expand on :options
+  def start_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> Temporal::WorkflowHandle
+
+  # Start the workflow and block awaiting it's completion
+  def execute_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> Any
+
+  # Generate a handle object from a namespace, workflow_id and a run_id
+  def workflow_handle(namespace, workflow_id, run_id = nil) -> Temporal::WorkflowHandle
+end
+```
+
+Notes:
+
+- This is not a complete definition of the client
+- Both positional arguments and keyword arguments are supported, except for the reserved `options:`
+- Conventions similar to Python SDK are used here:
+  - `Client` is a general purpose interface, not limited workflow-only interactions
+  - Due to this method names are more descriptive than just `#start`, `#execute`, etc
+  - `Client#execute_workflow` is added for convenience
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a workflow_id/run_id combination using the
+`Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class Temporal::WorkflowHandle
+  # Return the result of the workflow execution or raise if a workflow is still running after
+  #   waiting for for up to :timeout seconds.
+  # Passing timeout: nil will not wait if the working is still running
+  def result(timeout: Integer) -> Any
+
+  # Returns more information about a workflow
+  def describe() -> Temporal::WorkflowExecutionInfo
+
+  # Sends a cancellation request
+  def cancel() -> void
+
+  # Sends a query and awaits for the response
+  def query(name, *args) -> Any
+
+  # Sends a signal
+  def signal(name, *args) -> void
+
+  # Terminates a workflow
+  def terminate(reason, details: nil) -> void
+end
+```
+
+### Alternatives considered
+
+We thought about configuring the `Temporal::Client` with a namespace, similar to the Python SDK.
+It does bring the benefit of avoiding namespace to subsequent calls to the client. However we feel
+like it is an unjustified and artificial limitation that will impact some multi-tenant applications.
+Besides this can always be brought in the shape of a `NamespacedClient` if there's a need for it.
+
+
+## Worker
+
+The worker interface allows SDK users to start processing registered workflow and activities. These
+will then be polled for on the specified namespaces & task queues and executed.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+
+worker_1 = Temporal::Worker.new(config, 'my-namespace', 'my-task-queue')
```

Comment:
How do I have multiple workers reuse a client?

Reply:
> They reuse a connection that is generated on a configuration.

Ok, I think this point is definitely worth making clear. So where is the underlying gRPC client (that I suppose will come from core) stored? Is it literally stored on the configuration? Does it have to be that exact configuration instance to share the connection or are you going to perform some kind of hashing of the configuration? Hopefully not the latter as that can get confusing quick.

It is also confusing to a user that something as innocuous as reusing what may seem like simple stateless configuration actually affects connection count.

Reply:
Yes, I'm planning to store a gRPC client from Core on a configuration instance as mentioned here — https://github.com/temporalio/proposals/pull/57/files#diff-67ce557f10b281f1e3425a72e240ee009ad042859cdee9c99f00cd339bac6e3dR376

Reply:
That might be a bit surprising that configuration is more than configuration. I am trying to think of a better name/approach to the gRPC client that can be reused:

* `Configuration` - implies it's nothing more than just the fields/options you set and is surprising that it affects connection reuse (i.e. it's non-obvious that recreating a configuration with the same values has real connection reuse imlplications)
* `Client` - This is already taken to mean something else in this SDK :-( In some other SDKs including Python and Go, the configuration is just configuration and the client is the gRPC client, which makes pretty good sense when you think about it
* `Connection` - But then you have `Configuration` -> `Connection` -> `Client` which means we now have three levels of things instead of two

Reply:
@cretz but in Go and Python — client is not a gRPC client. It's a layer on top of a gRPC client, isn't it? Otherwise I would expect strict proto-defined interface from it.

I see `Configuration` as a vessel to pass `Connection` between a `Client` and a `Worker`. I agree that using `Configuration` for this is not ideal (you don't expect it on the configuration object), however it's a trade-off similar to having to initialize a `Client` to pass `Connection` (along with other configuration) to a `Worker` (which has nothing to do with the `Client`, it just needs a `Connection`)